### PR TITLE
Update status validation for ds/deploy/rs/rc

### PR DESCRIPTION
**Pull request by 0xmichalis**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36243)

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -2681,6 +2681,19 @@ func ValidateReplicationControllerStatusUpdate(controller, oldController *api.Re
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(controller.Status.ReadyReplicas), statusPath.Child("readyReplicas"))...)
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(controller.Status.AvailableReplicas), statusPath.Child("availableReplicas"))...)
 	allErrs = append(allErrs, ValidateNonnegativeField(int64(controller.Status.ObservedGeneration), statusPath.Child("observedGeneration"))...)
+	msg := "cannot be greater than status.replicas"
+	if controller.Status.Replicas < controller.Status.FullyLabeledReplicas {
```
Comment (lavalamp) [inline pkg/api/validation/validation.go]: tiny nit, but I would put the conditional in the same order as the message to save readers from mentally inverting things.

File: pkg/apis/extensions/validation/validation.go [hunk 2]
```diff
@@ -276,6 +277,16 @@ func ValidateDeploymentStatus(status *extensions.DeploymentStatus, fldPath *fiel
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.UpdatedReplicas), fldPath.Child("updatedReplicas"))...)
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.AvailableReplicas), fldPath.Child("availableReplicas"))...)
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.UnavailableReplicas), fldPath.Child("unavailableReplicas"))...)
+	msg := "cannot be greater than status.replicas"
+	//if status.Replicas < status.ReadyReplicas {
```
Comment (lavalamp) [inline pkg/apis/extensions/validation/validation.go]: delete dead code (or add comment with justification)

Below, too.

**Comment by lavalamp** [comment]
changes like this make me a bit nervous-- are there objects in systems right now that are in violation of these guidelines? If so we need to fix whatever controller is broken before making this change, or said controllers will hammer their apiservers.

**Comment by kargakis** [comment]
> changes like this make me a bit nervous-- are there objects in systems right now that are in violation of these guidelines? If so we need to fix whatever controller is broken before making this change, or said controllers will hammer their apiservers.

No, currently they are not (to my knowledge). Eventually, giving administrators the power to update objects status or adding more components like custom controllers will require us to tighten status validation.

**Comment by kargakis** [comment]
@lavalamp [here's](https://github.com/kubernetes/kubernetes/blob/bbd447f555380a6e41751d90c9f25d9388f939ab/pkg/controller/replicaset/replica_set_utils.go#L110-L120) how we calculate the status for rs/rc. Basically len(filteredPods) is status.replicas. Everything else should be at most len(filteredPods) so we are not broken by this extra validation. Deployments aggregate replica set status so it should be the same for them.

**Comment by lavalamp** [comment]
/lgtm

sorry for delay

**Comment by kargakis** [comment]
@k8s-bot kops e2e test this

**Comment by kargakis** [comment]
@k8s-bot unit test this https://github.com/kubernetes/kubernetes/issues/39604

**Pull request merged**